### PR TITLE
[LBSE] Add support for LBSE masks in SVGResourceImage

### PR DESCRIPTION
**Pull request by rwlbuis**
#### 6999be889df2d2b5b9ba7b5c92cd35f7b0ee848a
<pre>
[LBSE] Add support for LBSE masks in SVGResourceImage
<a href="https://bugs.webkit.org/show_bug.cgi?id=272823">https://bugs.webkit.org/show_bug.cgi?id=272823</a>

Reviewed by Nikolas Zimmermann.

Add support for LBSE masks in SVGResourceImage and
that way support HTML referencing SVG masks, both
in same document and in external SVG documents.

* LayoutTests/TestExpectations:
* Source/WebCore/rendering/ReferencedSVGResources.cpp:
(WebCore::ReferencedSVGResources::referencedMaskElement):
* Source/WebCore/rendering/ReferencedSVGResources.h:
* Source/WebCore/rendering/style/StyleCachedImage.cpp:
(WebCore::StyleCachedImage::legacyRenderSVGResource const):
(WebCore::StyleCachedImage::renderSVGResource const):
(WebCore::StyleCachedImage::isRenderSVGResource const):
(WebCore::StyleCachedImage::image const):
* Source/WebCore/rendering/style/StyleCachedImage.h:
* Source/WebCore/rendering/svg/RenderSVGResourceMasker.cpp:
(WebCore::RenderSVGResourceMasker::applyMask):
(WebCore::RenderSVGResourceMasker::drawContentIntoContext):
* Source/WebCore/rendering/svg/RenderSVGResourceMasker.h:
* Source/WebCore/svg/graphics/SVGResourceImage.cpp:
(WebCore::SVGResourceImage::create):
(WebCore::SVGResourceImage::SVGResourceImage):
(WebCore::SVGResourceImage::draw):
* Source/WebCore/svg/graphics/SVGResourceImage.h:

Canonical link: <a href="https://commits.webkit.org/278102@main">https://commits.webkit.org/278102@main</a>
</pre>

https://github.com/WebKit/WebKit/commit/0ea0548578483f7d840251897b9ca6abde9a8f1c

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/49498 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/28780 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/52536 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/52737 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/171 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/51802 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/34798 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/26400 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/40405 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/51598 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/26317 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/42641 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/21523 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/23777 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/43824 "Passed tests") | [✅ 🛠 wpe-skia](https://ews-build.webkit.org/#/builders/52/builds/7867 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/45691 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/44330 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/54249 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/24581 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/20759 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/47774 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/25852 "Built successfully") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/8/builds/42824 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/46795 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/10871 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/26692 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/25575 "Built successfully") | | |

### CODE REVIEW

File: Source/WebCore/rendering/ReferencedSVGResources.h [hunk 1]
```diff
@@ -74,6 +74,7 @@ class ReferencedSVGResources {
     static RefPtr<SVGClipPathElement> referencedClipPathElement(TreeScope&, const ReferencePathOperation&);
     static RefPtr<SVGMarkerElement> referencedMarkerElement(TreeScope&, const String&);
     static RefPtr<SVGMaskElement> referencedMaskElement(TreeScope&, const StyleImage&);
+    static RefPtr<SVGMaskElement> referencedMaskElement(TreeScope&, const AtomString& fragment);
```
Comment (nikolaszimmermann) [inline Source/WebCore/rendering/ReferencedSVGResources.h]: You can omit the parameter description, no?

File: Source/WebCore/rendering/style/StyleCachedImage.cpp [hunk 2]
```diff
@@ -278,6 +318,9 @@ RefPtr<Image> StyleCachedImage::image(const RenderElement* renderer, const Float
     if (auto renderSVGResource = this->renderSVGResource(renderer))
         return SVGResourceImage::create(*renderSVGResource, reresolvedURL(renderer->document()));
 
+    if (auto renderSVGResource = this->legacyRenderSVGResource(renderer))
```
Comment (nikolaszimmermann) [inline Source/WebCore/rendering/style/StyleCachedImage.cpp:321]: This adds new code for legacy? Why is this necessary?

File: Source/WebCore/rendering/svg/RenderSVGResourceMasker.cpp [hunk 3]
```diff
@@ -186,4 +179,35 @@ void RenderSVGResourceMasker::removeReferencingCSSClient(const RenderElement& cl
...
+
+    // Draw the content into the ImageBuffer.
+    checkedLayer()->paintSVGResourceLayer(context, maskContentTransformation);
+    return true;
+}
+
+bool RenderSVGResourceMasker::drawContentIntoContext(GraphicsContext& context, const FloatRect& destinationRect, const FloatRect& sourceRect, ImagePaintingOptions options)
+{
+    GraphicsContextStateSaver stateSaver(context);
+
```
Comment (nikolaszimmermann) [inline Source/WebCore/rendering/svg/RenderSVGResourceMasker.cpp]: I think you can omit some newlines here for readability.

**Review by nikolaszimmermann — APPROVED** [review]
Good catch, how could we miss that? 😅

File: Source/WebCore/rendering/svg/RenderSVGResourceMasker.cpp (hunk 3 above)
Comment (rwlbuis) [reply Source/WebCore/rendering/svg/RenderSVGResourceMasker.cpp]: Fair enough, removed some.

File: Source/WebCore/rendering/ReferencedSVGResources.h (hunk 1 above)
Comment (rwlbuis) [reply Source/WebCore/rendering/ReferencedSVGResources.h]: Done.

File: Source/WebCore/rendering/style/StyleCachedImage.cpp (hunk 2 above)
Comment (rwlbuis) [reply Source/WebCore/rendering/style/StyleCachedImage.cpp:321]: It is not adding legacy code, this is just a renaming. The new code added is for LBSE.

Comment (rwlbuis) [reply Source/WebCore/rendering/style/StyleCachedImage.cpp:321]: Note, since the aim is to remove legacy, I did not try any fancy solutions like templates and stuff, so things look a bit duplicated, but I think it is okay for now.

**Comment by rwlbuis** [comment]
> Good catch, how could we miss that? 😅

I was surprised too! I think at least partly it is because we/I focused a lot on the internal tests rather than WPT tests. Also since the masking tests do not fall under a svg/ parent dir, they are more easily forgotten about IMHO.
